[pkg/ottl] Add support for optional parameters

evan-bradley opened the pull request:
**Description:**

Provides one possible path for allowing functions to define optional parameters in OTTL.

This uses Python-style keyword arguments, so function calls will look like `ConvertCase(attributes["source"], delimiter = ".")`. Proposed rules for arguments after this change:
1.  camelCase/snake_case versions of the struct field names for the parameter can be used in function calls to specify an argument.
2. All named arguments must come after unnamed arguments.
3. Unnamed arguments must be in function signature order, named arguments can be in any order.
4. Arguments can be passed in unnamed or named.

**Link to tracking Issue:**

https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/20879

evan-bradley committed:
Add support for optional parameters in OTTL

evan-bradley commented:
I want to call out that the the diff for this is currently very much a work in progress. The changes to the grammar will require a huge set of changes to anything that reads arguments from the AST, and I left out any form of code cleanliness or testing to quickly prototype the idea.

TylerHelmuth commented:
> The changes to the grammar will require a huge set of changes to anything that reads arguments from the AST

This is such an annoying issue with OTTL and I wish I had some time to spend on it

evan-bradley commented:
>This is such an annoying issue with OTTL and I wish I had some time to spend on it

Unfortunately, since this is changing the parser I'm not sure there's a way around it. I'm with you there though, would be nice to fix it where we can.

evan-bradley commented on pkg/ottl/functions.go:
```diff
@@ -48,12 +48,16 @@ func (p *Parser[K]) newFunctionCall(ed editor) (Expr[K], error) {
 }
 
 func (p *Parser[K]) buildArgs(ed editor, argsVal reflect.Value) error {
+	// TODO: Determine number of optional args for this check.
```
I updated the PR description to outline the rules I'm thinking of setting for named and optional args. Let me know how it sounds.

evan-bradley commented on pkg/ottl/parser_test.go:
```diff
@@ -1462,7 +1462,7 @@ func Test_parseStatement(t *testing.T) {
 		{`set() where false or 1 == int() `, true},
 		{`set(foo.attributes["bar"].cat, "dog")`, false},
-		{`set(foo.attributes["animal"], "dog") where animal == "cat"`, false},
+		{`set(set = foo.attributes["animal"], val = "dog") where animal == "cat"`, false},
```
This test just checks that the parsing library can parse the grammar, but I've made an edit to `replace_pattern` to use it as the example function for this testing. I've tested the following statements:
- `replace_pattern(attributes["device"], ".*", "$$1")`
- `replace_pattern(attributes["device"], ".*", "$$1", function = "SHA256")`
- `replace_pattern(attributes["device"], ".*", "$$1", "SHA256")`

rnishtala-sumo replied:
Could we also add a test that demonstrates passing in an optional param and otherwise? I think that test would be in `function_test.go`

evan-bradley replied:
I'm trying to keep tests light since this is PR is still undergoing heavy revisions, but I'll make sure to add that before I mark the PR as ready for a proper review.

evan-bradley commented on pkg/ottl/grammar.go:
```diff
@@ -211,9 +219,19 @@ func (i *editor) checkForCustomError() error {
+	Keys      []Key      `parser:"( @@ )*"`
+}
+
+type argument struct {
```
Sorry, my naming wasn't particularly descriptive to make it clear what this does. `argument` will be used to hold arguments in place of doing it directly through `value`. This allows us to associate a name with the argument.

rnishtala-sumo replied:
ah ok, just to confirm are named arguments going to be considered optional, or are they two different things? It looks like the grammar would support named arguments and the parser treats them as optional when they're passed in?

evan-bradley commented on pkg/ottl/functions.go:
```diff
@@ -48,12 +48,16 @@ func (p *Parser[K]) newFunctionCall(ed editor) (Expr[K], error) {
 
 	argsType := argsVal.Type()
 
+	// TODO: Handle named arguments
```
I'm trying to keep all the arguments in one struct, and allow any arguments to be named. See [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/24456/files#diff-d32daaa4e1064c4af10a5eb364737ef2bdff7c94d15dbd94a796ddb42ec2afa8R14) for an example of what a definition might look like.

rnishtala-sumo commented on pkg/ottl/functions.go:
```diff
@@ -70,13 +97,60 @@ func (p *Parser[K]) buildArgs(ed editor, argsVal reflect.Value) error {
+			return fmt.Errorf("invalid argument at position %v: %w", i, err)
 		}
+		if isOptional {
+			fieldVal.Set(reflect.ValueOf(val))
```
what happens when an optional param is not passed in, would this be `nil`?

evan-bradley replied:
The `ottl.Optional` type will indicate (currently through `HasValue` or `IsEmpty`) whether a value is present, but the field value here will be the zero value of the underlying type, which usually won't be `nil`. The existence of zero values in Go necessitates having another way to indicate whether a value was passed in. We could potentiall use pointers here (which have `nil` as a zero value), but I would prefer to have something that is more explicit.

rnishtala-sumo commented on pkg/ottl/functions.go:
```diff
@@ -86,7 +160,17 @@ func (p *Parser[K]) buildArgs(ed editor, argsVal reflect.Value) error {
 		}
-		field.Set(reflect.ValueOf(val))
+		if isOptional {
+			fieldVal.Set(reflect.ValueOf(val))
```
>what happens when an optional param is not passed in, would this be nil?

Same question here.

evan-bradley replied:
`presenceVal` is maybe a poor name, something like `precenseField` would be more accurate. This is the field on the `Optional` struct that indicates whether a value has been set. By default it is `false`, indicating that the user has not passed in a value for this optional parameter. If the user does pass something in, we set it to `true`.

evan-bradley commented:
@rnishtala-sumo @TylerHelmuth I believe there's enough here now for a very high-level review of the approach.

For optional parameters, I want three features:
1. Functions can tell whether an optional parameter was passed a value by the user
2. Function authors can easily specify default values for optional parameters
3. Optional parameters can be specified independently of each other

Feature 2 isn't a hard requirement since it can be fulfilled by 1.

I've implemented features 1 and 2 using an `Optional` type. Feature 1 would not be possible with plain types due to Go's use of zero-values.

Feature 3 is achieved using named arguments that work like Python's keyword arguments. A function call with a named argument would look like `ConvertCase(attributes["source"], delimiter = ".")`. Proposed rules for arguments after this change:
 1. camelCase/snake_case versions of the struct field names for the parameter can be used in function calls to specify an argument. We can enforce the casing if we like, but I think anything that can be converted to PascalCase (which is the casing for a public struct field) will work. I'll also add a struct tag to allow function authors to specify the name if they want something custom.
 2. All named arguments must come after any unnamed arguments.
 3. Unnamed arguments must be in function signature order, while named arguments can be in any order.
 4. Arguments can be passed in unnamed or named, including optional arguments.

I believe I've implemented all of the optional parameter features and named argument features on a basic level. The current implementation is just to test the viability of the approach and give a general sense of what the final changes will look like. It's still in a rough state, so if anything outside of the high-level design seems off, assume it will be fixed.

Issues I am still working through:
1. The way we create factories means that a single instance of each factory's `Arguments` struct is used for all function instantiations. This wasn't previously an issue, but it needs to be fixed now.
2. I don't like that `Optional` is a struct with public fields, but the way that Go reflection and generics work, it has been the only viable approach I've found so far. My ultimate goal would be to make it an interface with `IsEmpty`/`HasValue` and `Get` methods, but barring that, I have a few ideas for cleaning this up that I will test.
3. Ensuring argument validation is thorough given the changes. I'll do this before marking the PR ready for review.

TylerHelmuth commented on pkg/ottl/grammar.go:
```diff
@@ -211,9 +220,20 @@ func (i *editor) checkForCustomError() error {
+}
+
+type argument struct {
+	Name  string `parser:"(@(Lowercase(Uppercase | Lowercase)*) Equal)?"`
```
I just realized our tokens allow for function, and now parameter, names like `1` or `0`.  We should probably not allow that, but we can fix that in a separate PR.

TylerHelmuth commented on pkg/ottl/functions.go:
```diff
@@ -48,15 +50,40 @@ func (p *Parser[K]) newFunctionCall(ed editor) (Expr[K], error) {
+	namedArgs := []argument{}
+	i := 0
+
+	for ; i < len(ed.NewArguments); i++ {
+		if ed.NewArguments[i].Name == "" {
+			unnamedArgs = append(unnamedArgs, ed.NewArguments[i])
+		} else {
+			break
+		}
+	}
+	for ; i < len(ed.NewArguments); i++ {
+		if ed.NewArguments[i].Name != "" {
+			namedArgs = append(namedArgs, ed.NewArguments[i])
+		} else {
+			return errors.New("Unnamed argument used after named argument")
+		}
+	}
```
We could handled this setup and error in a single loop if we use a bool to keep track of when a named argument has been seen.

evan-bradley replied:
Done.

TylerHelmuth commented on pkg/ottl/functions.go:
```diff
@@ -291,3 +375,27 @@ func buildSlice[T any](argVal value, argType reflect.Type, buildArg buildArgFunc
+	}
+}
+
+type Optional[T any] struct {
```
I agree that an interface would be nice, but as a convenience for function creator's we'll need a `StandardOptional` struct anyways.

evan-bradley replied:
Could you elaborate? I'm having trouble using an interface right instead of a struct because the parser can't use the zero value for an interface and I can't figure out how to have the parser fix that on its own. We would need to have it create a `StandardOptional[T]` struct with the necessary data, then set it on the `Optional[T]` field in the `Arguments` struct. So for now we're stuck with just an `Optional` struct.

TylerHelmuth replied:
I have no solution at the moment, only the theoretical concept that an interface would be great lol Is there any way to unexport the fields and only expose getting and setting them via functions that we execute with reflection? That would make the exported Optional type easier to modify in the feature without breaking changes to the API

evan-bradley replied:
Unfortunately we can't call private methods with reflection, but we can work around the reflection limitations by having `Optional[T]` implement an interface that only deals with `any` and `reflect.Value` types. It's all a very delicate puzzle, and for now I think I've exhausted the options I'm aware of outside of meddling with `unsafe.Pointer` types.

evan-bradley replied:
I'm still unable to think of a way to use an interface, but an `Optional` struct will probably meet all our needs. If you think of a reason why we will need an interface I will take another look at whether it's feasible.

TylerHelmuth replied:
I'm ok moving forward with the Optional struct

TylerHelmuth commented on pkg/ottl/functions.go:
```diff
@@ -70,13 +97,60 @@ func (p *Parser[K]) buildArgs(ed editor, argsVal reflect.Value) error {
 		}
+		if isOptional {
+			fieldVal.Set(reflect.ValueOf(val))
+			presenceVal := field.FieldByName("HasValue")
+			presenceVal.Set(reflect.ValueOf(true))
```
Instead of reflection is there any way to work with the Optional struct directly since we know that is what we have?

evan-bradley replied:
I ended up going with the interface and it turns out I like it. Please take a look.

TylerHelmuth commented on pkg/ottl/functions.go:
```diff
@@ -70,13 +97,60 @@ func (p *Parser[K]) buildArgs(ed editor, argsVal reflect.Value) error {
+	}
 
-		argVal := ed.Arguments[argNum]
+	for i := 0; i < len(namedArgs); i++ {
```
I'm missing something, is this loop the same functionality as the one above?

rnishtala-sumo replied:
I believe we have different loops for named and unnamed args. The only difference seems to be in how the `field` variable is set.

evan-bradley replied:
This was mostly for prototyping, I've merged the loops now.

rnishtala-sumo commented:
> Ensuring argument validation is thorough given the changes. I'll do this before marking the PR ready for review.

@evan-bradley is this the only remaining item in this PR? I'm referring to this [comment](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/24456#issuecomment-1664333423)

TylerHelmuth commented:
@evan-bradley set this as Ready for Review the next time you want me to take a look, otherwise I'm gonna assume you're still ideating

evan-bradley commented:
This should be ready for a review.

I removed support for supplying default values for Optional parameters since it would be too burdensome to create default Getter values and using `IsEmpty` enables that without much trouble. I can revisit that if there's a desire to support defaults. The downside is that now the name of the `CreateDefaultArguments` method on the `Factory` interface makes less sense if it doesn't provide defaults.

TylerHelmuth reviewed (commented):
@evan-bradley I've done a pass through this, but would like to merge https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/25705 first as I believe it will simplify this PR.

evan-bradley commented:
@TylerHelmuth This should be ready for another look, I've rebased on the changes to remove the struct tag.

TylerHelmuth reviewed (approved):
LGTM, left 1 non-blocking question. Also, this PR makes me wish we had some e2e tests in OTTL itself instead of relying on the transformprocessor.

evan-bradley commented on pkg/ottl/functions.go:
```diff
@@ -310,3 +365,56 @@ func buildSlice[T any](argVal value, argType reflect.Type, buildArg buildArgFunc
+// nolint:unused
+func (o Optional[T]) set(val any) reflect.Value {
+	return reflect.ValueOf(Optional[T]{
+		val:      val.(T),
```
I thought about this for a bit and I don't think it's really possible to hit a case where this fails. We build `val` based on `T` before calling this, so they will always line up so long as `buildArg` is working correctly. We make essentially the same assertion without using type conversion [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/24456/files#diff-7d00262f5fd3b4cb4b569750e059f83615378f288c84266f86b59b3747143f71R138) for required parameters.
